### PR TITLE
Feature/lattice sampling rework

### PR DESCRIPTION
**Pull request by yangmr04**
### `AbstractWalkers` module:
- Refactored `LatticeSystem{G}` to be parametrized by `G`, a new abstract type `LatticeGeometry` to reflect the geometry of the lattice. Subtypes of `G` are `SquareLattice`, `TriangularLattice` and `GenericLattice`.

### `AbstractLiveSets` module
- Added `LatticeGasWalkers <: LatticeWalkers` for lattice nested sampling, etc.

### `EnergyEval` module
- Energies are now evaluated with function `interacting_energy(lattice::LatticeSystem, h::GenericLatticeHamiltonian{N})`, which is consistent with how atomistic systems are handled.

### `Hamiltonians` module
- Added `GenericLatticeHamiltonian{N,U} <: ClassicalHamiltonian` where `N` is the furtherest neighbor and `U` is the units of energy.

### `AtomsMCMoves` module:
- Refactored `AtomsMCMoves` module into `MonteCarloMoves` to accommodate lattice MC moves. Split off the normal MC moves and NVE moves.

### `SamplingSchemes` module:
- Separated nested sampling, Wang-Landau, NVT MC and experimental replica exchange into independent files.
- All lattice sampling schemes are now using unified `sampling_scheme(lattice::LatticeSystem, h:: ClassicalHamiltonian; args...)` style. 
- Added multiple dispatches for lattice nested sampling functions, now the call signatures for nested sampling functions are consistent for both lattice and atomistic systems, e.g., 
```
nested_sampling_loop!(liveset::AbstractLiveSet,
                      ns_params::SamplingParameters, 
                      n_steps::Int64, 
                      mc_routine::MCRoutine,
                      save_strategy::DataSavingStrategy)
```
- Made `exact_enumeration`, `nvt_monte_carlo`, `wang_landau` all compatible with new lattice walker types and works with `ClassicalHamiltonian`s.

### CODE REVIEW

File: scripts/example-mc.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Define parameters for MCMC simulation
+L = square_supercell_dimensions[1]
+M = square_supercell_dimensions[2]
+Ns = [4]
+adsorption_energy = -0.04
+nn_energy = -0.01
+nnn_energy = -0.0025
```
Comment (rwexler) [inline scripts/example-mc.jl:7]: There's a logic to these energies that should be reflected in the tutorial for MCMC in the documentation.

File: scripts/example-mc.jl [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+temperatures = 200:-1:1
+num_equilibration_steps = 10_000
+num_sampling_steps = 10_000
+random_seed = 1234
+k_B = 8.617_333_262e-5
```
Comment (rwexler) [inline scripts/example-mc.jl:12]: @yangmr04: Can we hard-code this (NIST's) value for k_B (in eV/K) into FreeBird.jl?

File: scripts/example-mc.jl [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+# Define file paths for data
+parameters = "mc_square_L
+(M)_adsorption
+(nn_energy)_nnn
+(square_cutoff_radii[1])_$(square_cutoff_radii[2])"
```
Comment (rwexler) [inline scripts/example-mc.jl:18]: This file path looks strange. @yangmr04: Can you confirm that this is correct?

File: scripts/example-mc.jl [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+
+df_samples = DataFrame(N = Int[], T = Float64[], E = Float64[], Cv = Float64[], acceptance_rate = Float64[])
+
+# Perform the Monte Carlo simulation
+for N in Ns
+    println("N = $N -------------------")
+
+    # Initialize the lattice
```
Comment (rwexler) [inline scripts/example-mc.jl:27]: Do we have any convenience functions for lattice initialization yet?

File: scripts/example-mc.jl (hunk 2 above)
Comment (yangmr04) [reply scripts/example-mc.jl:12]: Yup, can be done by declaring a constant with `const k_B` in FreeBird.

File: scripts/example-mc.jl (hunk 4 above)
Comment (yangmr04) [reply scripts/example-mc.jl:27]: Getting there! Adding more!

File: scripts/example-mc.jl (hunk 3 above)
Comment (yangmr04) [reply scripts/example-mc.jl:18]: It works. But will be modified once turned into a tutorial.

**Pull request merged**